File: testing/orm-live/django/suite.py
#!/usr/bin/env python3
"""Live Django ORM suite against BASIN_DSN.

Drives Django's OWN machinery: makemigrations (offline) + migrate (its
migration executor and schema editor), then the ORM query compiler,
transaction.atomic with savepoints, F() expressions and bulk_create.

TAP-ish output: "ok - django.<test>" / "not ok - django.<test> # <reason>".
"""

import os
import subprocess
import sys
import time

HERE = os.path.dirname(os.path.abspath(__file__))
RUN = time.time_ns()


def ok(name):
    print(f"ok - django.{name}", flush=True)


def notok(name, reason):
    reason = " ".join(str(reason).split())[:300]
    print(f"not ok - django.{name} # {reason}", flush=True)


REMAINING = ["crud", "select-related", "atomic-commit", "savepoint-rollback",
             "f-update", "bulk-create"]

if not os.environ.get("BASIN_DSN"):
    notok("migrate", "BASIN_DSN not set")
    sys.exit(0)

# ── 1. Django's migration engine: makemigrations (offline) + migrate ────────
try:
    for args in (["makemigrations", "blog", "--noinput"], ["migrate", "--noinput"]):
        r = subprocess.run(
            [sys.executable, os.path.join(HERE, "manage.py"), *args],
            capture_output=True, text=True, timeout=180, cwd=HERE,
        )
        if r.returncode != 0:
            raise RuntimeError((r.stderr or r.stdout).strip()[-300:])
    ok("migrate")
    migrate_ok = True
except Exception as e:  # noqa: BLE001
    notok("migrate", e)
    migrate_ok = False

# ── ORM session ──────────────────────────────────────────────────────────────
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "basinsite.settings")
sys.path.insert(0, HERE)
import django  # noqa: E402

django.setup()

from django.db import connection, transaction  # noqa: E402
from django.db.models import F  # noqa: E402
from blog.models import Author, Book  # noqa: E402

try:
    with connection.cursor() as cur:
        cur.execute("SELECT 1")
except Exception as e:  # noqa: BLE001
    notok("connect", e)
    for t in REMAINING:
        notok(t, "SKIP-CASCADE: no connection")
    sys.exit(0)
else:
    ok("connect")

if not migrate_ok:
    # Tables may still exist from a previous run; probe before cascading.
    try:
        Author.objects.exists()
    except Exception:  # noqa: BLE001
        for t in REMAINING:
            notok(t, "SKIP-CASCADE: migrate failed and tables absent")
        sys.exit(0)


def test(name, fn):
    try:
        fn()
        ok(name)
    except Exception as e:  # noqa: BLE001
        notok(name, e)


def check(cond, msg):
    if not cond:
        raise AssertionError(f"assertion failed: {msg}")


def t_crud():
    a = Author.objects.create(name=f"alice-{RUN}")
    check(a.pk and a.created_at is not None, "create hydrates pk + auto_now_add")
    b = Book.objects.create(author=a, title="One", pages=100, meta={"genre": "db"})
    got = Book.objects.get(pk=b.pk)
    check(got.meta == {"genre": "db"}, "JSONField round-trips")
    got.pages = 110
    got.save()
    check(Book.objects.get(pk=b.pk).pages == 110, "update persisted")
    victim = Book.objects.create(author=a, title="Victim")
    victim.delete()
    check(not Book.objects.filter(title="Victim", author=a).exists(), "delete persisted")


def t_select_related():
    book = Book.objects.select_related("author").get(title="One", author__name=f"alice-{RUN}")
    check(book.author.name == f"alice-{RUN}", "select_related JOIN hydrates the FK")


def t_atomic_commit():
    with transaction.atomic():
        a = Author.objects.create(name=f"tx-{RUN}")
        Book.objects.create(author=a, title="InTx")
    check(Author.objects.filter(name=f"tx-{RUN}").exists(), "committed atomic block visible")


def t_savepoint_rollback():
    with transaction.atomic():
        Author.objects.create(name=f"outer-{RUN}")
        try:
            with transaction.atomic():  # inner block -> SAVEPOINT
                Author.objects.create(name=f"inner-{RUN}")
                raise RuntimeError("force inner rollback")
        except RuntimeError:
            pass
    check(Author.objects.filter(name=f"outer-{RUN}").exists(), "outer write committed")
    check(not Author.objects.filter(name=f"inner-{RUN}").exists(),
          "inner write rolled back to savepoint")


def t_f_update():
    n = Book.objects.filter(title="One", author__name=f"alice-{RUN}").update(pages=F("pages") + 5)
    check(n == 1, "F() update affected exactly one row")
    check(Book.objects.get(title="One", author__name=f"alice-{RUN}").pages == 115,
          "F() arithmetic applied server-side (110 + 5)")


def t_bulk_create():
    objs = Author.objects.bulk_create([Author(name=f"bulk-{RUN}-{i}") for i in range(5)])
    check(len(objs) == 5, "bulk_create returned all objects")
    check(Author.objects.filter(name__startswith=f"bulk-{RUN}-").count() == 5,
          "all bulk rows persisted")
    # ignore_conflicts path (INSERT ... ON CONFLICT DO NOTHING).
    Author.objects.bulk_create(
        [Author(name=f"bulk-{RUN}-0")], ignore_conflicts=True)
    check(Author.objects.filter(name=f"bulk-{RUN}-0").count() == 1,
          "ignore_conflicts did not duplicate")


test("crud", t_crud)
test("select-related", t_select_related)
test("atomic-commit", t_atomic_commit)
test("savepoint-rollback", t_savepoint_rollback)
test("f-update", t_f_update)
test("bulk-create", t_bulk_create)

File: testing/orm-live/typeorm/suite.mjs
// Live TypeORM suite — DataSource init, TypeORM's own migration runner,
// repository CRUD, eager relations, and QueryRunner-managed transactions,
// against BASIN_DSN.
// TAP-ish output: "ok - typeorm.<test>" / "not ok - typeorm.<test> # <reason>".

import { dataSource } from "./datasource.js";

const RUN = Date.now();

function ok(name) { console.log(`ok - typeorm.${name}`); }
function notok(name, reason) {
  console.log(`not ok - typeorm.${name} # ${String(reason ?? "unknown").replace(/\s+/g, " ").slice(0, 300)}`);
}

const REMAINING = ["migrations-run", "crud-insert", "crud-find", "crud-update",
                   "crud-delete", "eager-relations", "queryrunner-tx", "queryrunner-tx-rollback"];

if (!process.env.BASIN_DSN) {
  notok("connect", "BASIN_DSN not set");
  process.exit(0);
}

// ── 1. connect (DataSource.initialize) ───────────────────────────────────────
try {
  await dataSource.initialize();
  ok("connect");
} catch (e) {
  notok("connect", e.message ?? e);
  for (const t of REMAINING) notok(t, "SKIP-CASCADE: no connection");
  process.exit(0);
}

async function test(name, fn) {
  try { await fn(); ok(name); } catch (e) { notok(name, e.message ?? e); }
}
function assert(cond, msg) { if (!cond) throw new Error(`assertion failed: ${msg}`); }

// ── 2. TypeORM's migration runner (creates its "migrations" tracker table) ──
await test("migrations-run", async () => {
  const applied = await dataSource.runMigrations();
  // Idempotent: 1 on first run, 0 on re-run against a durable server.
  assert(Array.isArray(applied), "runMigrations returns applied list");
});

const users = dataSource.getRepository("User");
const posts = dataSource.getRepository("Post");
let alice;

// ── 3. Repository CRUD ───────────────────────────────────────────────────────
await test("crud-insert", async () => {
  alice = await users.save({ email: `alice.${RUN}@typeorm.test`, name: "Alice" });
  assert(alice.id > 0, "save() hydrated the generated id");
  await posts.save([
    { title: "t-one", views: 5, author: alice },
    { title: "t-two", author: alice },
  ]);
});

await test("crud-find", async () => {
  const found = await users.findOneBy({ email: `alice.${RUN}@typeorm.test` });
  assert(found && found.name === "Alice", "findOneBy returns the row");
  assert(found.createdAt instanceof Date, "createDate column hydrates");
});

await test("crud-update", async () => {
  await users.update({ email: `alice.${RUN}@typeorm.test` }, { name: "Alice Updated" });
  const found = await users.findOneBy({ email: `alice.${RUN}@typeorm.test` });
  assert(found.name === "Alice Updated", "update persisted");
});

await test("crud-delete", async () => {
  const victim = await users.save({ email: `victim.${RUN}@typeorm.test` });
  await users.delete({ id: victim.id });
  assert((await users.findOneBy({ id: victim.id })) === null, "deleted row gone");
});

// ── 4. Eager relations (Post.author is eager:true) + find w/ relations ──────
await test("eager-relations", async () => {
  const one = await posts.findOne({ where: { title: "t-one" } });
  assert(one && one.author && one.author.email === `alice.${RUN}@typeorm.test`,
    "eager many-to-one hydrated on plain find");
  const withPosts = await users.findOne({
    where: { id: alice.id },
    relations: { posts: true },
  });
  assert(withPosts.posts.length === 2, "one-to-many loaded via relations option");
});

// ── 5. QueryRunner-managed transactions ──────────────────────────────────────
await test("queryrunner-tx", async () => {
  const qr = dataSource.createQueryRunner();
  await qr.connect();
  await qr.startTransaction();
  try {
    await qr.manager.save("User", { email: `tx.${RUN}@typeorm.test` });
    await qr.commitTransaction();
  } catch (e) {
    await qr.rollbackTransaction();
    throw e;
  } finally {
    await qr.release();
  }
  const found = await users.findOneBy({ email: `tx.${RUN}@typeorm.test` });
  assert(found, "committed QueryRunner tx visible");
});

await test("queryrunner-tx-rollback", async () => {
  const qr = dataSource.createQueryRunner();
  await qr.connect();
  await qr.startTransaction();
  await qr.manager.save("User", { email: `rb.${RUN}@typeorm.test` });
  await qr.rollbackTransaction();
  await qr.release();
  const found = await users.findOneBy({ email: `rb.${RUN}@typeorm.test` });
  assert(found === null, "rolled-back QueryRunner write must not be visible");
});

await dataSource.destroy();
